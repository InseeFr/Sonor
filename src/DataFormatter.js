import Service from './Service.js'

class DataFormatter{

	static getDataForMainScreen(cb){
	  Service.getSurveys((data)=>{
	    cb(data)
	  });
	}

	static getDataForListSU(survey, cb){
	  Service.getSurveyUnits(survey, (res)=>{
	      const processedData=[]
	      res.forEach(su=>{
	        const suLine = {}
	        suLine.id = su.id
	        suLine.ssech = su.ssech
	        suLine.departement = su.location
	        suLine.city = su.city
	        suLine.interviewer = su.interviewer.first_name + ' ' + su.interviewer.last_name
	        suLine.idep = su.interviewer.idep
	        processedData.push(suLine)
	      })
	      cb(processedData)
	  })
	}

	static getDataForMonitoringTable(survey, cb){
	  const p1 = new Promise((resolve, reject) => {
	    Service.getInterviewers(survey, (res)=>{
	      const promises = []
	      res.forEach(interviewer=>{
	          promises.push(
	              new Promise((resolve2, reject2) => {
	                Service.getInterviewersStateCount(survey, interviewer.idep,(data)=>{resolve2({interviewer: interviewer, state_count: data})})
	              })
	            )
	      })
	      Promise.all(promises).then(data=>{
	        const processedData=[]
	        data.forEach(data=>{
	          const line = {}
	          line.interviewer = data.interviewer.interviewer_first_name + ' ' + data.interviewer.interviewer_last_name
	          formatForMonitoringTable(line, data.state_count)
	          processedData.push(line)
	        })
	        resolve(processedData)
	      })
	    })
	  });

	  const p2 = new Promise((resolve, reject) => {
	    Service.getStateCount(survey, (data)=>{
	      const totalDem = {}
	      formatForMonitoringTable(totalDem, data.DEM)

	      const totalFrance= {}
	      formatForMonitoringTable(totalFrance, data.France)


	      resolve({dem:totalDem, france:totalFrance})
	    })
	  })

	  Promise.all([p1,p2]).then(data=>{
	    cb({interviewers: data[0], total: data[1]})
	  });
	}


	static getDataForCampaignPortal(campaignId, cb){
		const p1 = new Promise((resolve, reject) => {Service.getInterviewersByCampaign(campaignId, 
		  (data)=>{resolve(data);});
		});
		const p2 = new Promise((resolve, reject) => {Service.getNotAttributedByCampaign(campaignId, 
		  (data)=>{resolve(data);});
		});
		const p3 = new Promise((resolve, reject) => {Service.getTotalDemByCampaign(campaignId, 
		  (data)=>{resolve(data);});
		});
		Promise.all([p1,p2,p3]).then(data=>{
		  cb({interviewers: data[0], notAttributed: data[1], total: data[2]})
		});
  	}

}

function formatForMonitoringTable(line, stateCount){
  line.completionRate = calculateCompletionRate(stateCount)
  line.total = stateCount.total
  line.notStarted = stateCount.ANS_count
  line.onGoing = calculateOngoing(stateCount)
  line.waitingForIntValidation = stateCount.WFT_count + stateCount.WFS_count
  line.intValidated = stateCount.TBR_count
  line.demValidated = stateCount.FIN_count
  line.preparingContact = stateCount.PRC_count
  line.atLeastOneContact = stateCount.AOC_count
  line.appointmentTaken = stateCount.APS_count
  line.interviewStarted = stateCount.INS_count
}

function calculateCompletionRate(data){
  return (data.total - data.INI_count - data.ANS_count)/data.total
}

function calculateOngoing(data){
  return data.PRC_count + data.AOC_count + data.APS_count + data.INS_count + data.WFT_count + data.WFS_count
}



export default DataFormatter;
