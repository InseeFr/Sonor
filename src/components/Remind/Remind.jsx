import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import Card from 'react-bootstrap/Card';
import Button from 'react-bootstrap/Button';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import SurveySelector from '../SurveySelector/SurveySelector';
import D from '../../i18n';

function Remind({
  location,
}) {
  const [survey, setSurvey] = useState(location.survey);
  const surveyTitle = !survey
      || (<div className="SurveyTitle">{survey.label}</div>);
  const surveySelector = !survey
      || (
        <SurveySelector
          survey={survey}
          updateFunc={(newSurvey) => setSurvey(newSurvey)}
        />
      );
  return (
    <div id="Remind">
      <Container fluid>
        <Row>
          <Col>
            <Link to="/" className="ButtonLink">
              <Button className="YellowButton ReturnButton" data-testid="return-button">{D.back}</Button>
            </Link>
          </Col>
          <Col xs={6}>
            {surveyTitle}
          </Col>
          <Col>
            {surveySelector}
          </Col>
        </Row>
      </Container>
      <Card className="ViewCard">
        <Card.Title>
          {D.developmentInProgress}
        </Card.Title>
      </Card>
    </div>
  );
}

export default Remind;
