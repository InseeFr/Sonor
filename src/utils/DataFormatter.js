import Service from './Service';
import Utils from './Utils';
import {
  BY_INTERVIEWER_ONE_SURVEY,
  BY_SURVEY,
  BY_SITE,
} from './constants.json';

class DataFormatter {
  constructor(token) {
    this.service = new Service(token);
  }

  getUserInfo(cb) {
    this.service.getUser((data) => {
      cb(data);
    });
  }

  getPreferences(cb) {
    return new Promise((resolve) => {
      this.service.getSurveys((data) => {
        const formattedData = {};
        data.forEach((survey) => {
          if (Utils.isVisible(survey)) {
            formattedData[survey.id] = {
              label: survey.label,
              preference: survey.preference,
            };
          }
        });
        if (cb) {
          cb(formattedData);
        }
        resolve(formattedData);
      });
    });
  }

  getDataForMainScreen(date, cb) {
    return new Promise((resolve) => {
      this.service.getSurveys((data) => {
        const formattedData = [];
        data.forEach((survey) => {
          if (Utils.isVisible(survey, date)) {
            const formattedSurvey = {};
            Object.assign(formattedSurvey, survey);
            formattedSurvey.phase = Utils.getCampaignPhase(
              survey.collectionStartDate,
              survey.collectionEndDate,
              survey.treatmentEndDate,
            );
            formattedData.push(formattedSurvey);
          }
        });
        if (cb) {
          cb(formattedData);
        }
        resolve(formattedData);
      });
    });
  }

  getDataForListSU(surveyId, cb) {
    const p1 = new Promise((resolve) => {
      this.service.getUser((res) => {
        resolve(res.organizationUnit.label);
      });
    });
    const p2 = new Promise((resolve) => {
      this.service.getSurveyUnits(surveyId, (res) => {
        const processedData = [];
        res.forEach((su) => {
          const suLine = {};
          suLine.id = su.id;
          suLine.ssech = su.ssech;
          suLine.departement = su.location;
          suLine.city = su.city;
          suLine.interviewer = `${su.interviewer.firstName} ${su.interviewer.lastName}`;
          suLine.idep = su.interviewer.id;
          processedData.push(suLine);
        });
        resolve(processedData);
      });
    });
    Promise.all([p1, p2]).then((data) => {
      cb({ site: data[0], surveyUnits: data[1] });
    });
  }

  getDataForReview(surveyId, cb) {
    this.getListSUToReview(surveyId).then((data) => {
      cb(data);
    });
  }

  finalizeSurveyUnits(suToFinalize, cb) {
    return new Promise((resolve) => {
      this.service.putSurveyUnitsToValidate(suToFinalize, (data) => {
        if (cb) {
          cb(data);
        }
        resolve(data);
      });
    });
  }

  updatePreferences(preferences, cb) {
    return new Promise((resolve) => {
      this.service.putPreferences(preferences, (data) => {
        if (cb) { cb(data); }
        resolve(data);
      });
    });
  }

  getListSUToReview(surveyId) {
    return new Promise((resolve) => {
      const promises = [];
      this.service.getSurveys((res) => {
        res.forEach((campaign) => {
          if (surveyId === null || campaign.id === surveyId) {
            promises.push(
              new Promise((resolve2) => {
                this.service.getSurveyUnits(campaign.id, (res2) => {
                  const lstSU = [];
                  res2.forEach((su) => {
                    const suLine = {};
                    suLine.campaignLabel = campaign.label;
                    suLine.interviewer = `${su.interviewer.lastName} ${su.interviewer.firstName}`;
                    suLine.idep = su.interviewer.id;
                    suLine.id = su.id;
                    lstSU.push(suLine);
                  });
                  lstSU.sort((a, b) => (a.interviewer > b.interviewer ? 1 : -1));
                  resolve2(lstSU);
                });
              }),
            );
          }
        });
        Promise.all(promises).then((data) => {
          resolve(data.flat());
        });
      });
    });
  }

  getInterviewersDetail(survey, interviewers, date, mode) {
    const surveyId = !survey || survey.id;
    return new Promise((resolve) => {
      const promises = [];
      interviewers.forEach((interv) => {
        promises.push(
          new Promise((resolve2) => {
            this.service.getInterviewersStateCount(
              surveyId || interv.survey,
              interv.id,
              date,
              (data) => {
                const interviewer = {};
                Object.assign(interviewer, interv);
                interviewer.survey = surveyId;
                resolve2({ interviewer, stateCount: data });
              },
            );
          }),
        );
      });
      if (mode === BY_SURVEY) {
        Promise.all(promises).then((data) => {
          const sum = Utils.getStateCountSum(data);
          sum.survey = survey.label;
          resolve(sum);
        });
      } else {
        Promise.all(promises).then((data) => {
          const processedData = [];
          data.forEach((intData) => {
            const line = {};
            Object.assign(line, intData);
            line.interviewerFirstName = intData.interviewer.interviewerFirstName;
            line.interviewerLastName = intData.interviewer.interviewerLastName;
            line.interviewerId = intData.interviewer.id;
            processedData.push(line);
          });
          resolve(processedData);
        });
      }
    });
  }

  async getDataForMonitoringTable(survey, date, pagination, mode, cb) {
    const interviewers = [];
    const getDataForSingleSurvey = !Array.isArray(survey);
    let p1;
    let site;

    if (mode !== BY_SITE) {
      let surveysToGetInterviewersFrom;
      if (getDataForSingleSurvey) {
        surveysToGetInterviewersFrom = [survey];
      } else {
        surveysToGetInterviewersFrom = await this.getDataForMainScreen(date);
      }
      site = (await this.service.getUser()).organizationUnit.label;
      p1 = new Promise((resolve) => {
        const promises = [];
        surveysToGetInterviewersFrom.forEach((surv) => {
          promises.push(
            new Promise((resolve2) => {
              this.service.getInterviewers(survey, (res) => {
                res.forEach((interviewer) => {
                  Utils.addIfNotAlreadyPresent(interviewers, interviewer);
                });
                this.getInterviewersDetail(
                  surv,
                  res,
                  date,
                  mode,
                ).then((data) => resolve2(data));
              });
            })
          );
        });
        Promise.all(promises).then((data) => {
          if (mode === BY_SURVEY) {
            resolve(data);
          } else {
            resolve(Utils.sumOn(data.flat(), 'interviewerId'));
          }
        });
      });
    }

    if (mode === BY_INTERVIEWER_ONE_SURVEY || mode === BY_SITE) {
      const p2 = new Promise((resolve) => {
        this.service.getStateCount(survey, date, (data) => {
          if (mode === BY_INTERVIEWER_ONE_SURVEY) {
            this.service.getUser((userInfo) => {
              const userOUs = userInfo.localOrganizationUnits.map((x) => x.id);
              const demStateCounts = [];
              data.organizationunits.forEach((dem) => {
                if (userOUs.includes(dem.idDem)) {
                  demStateCounts.push({ stateCount: dem });
                }
              });
              const totalDem = Utils.getStateCountSum(demStateCounts);
              const totalFrance = Utils.formatForMonitoringTable(data.France);
              resolve({ dem: totalDem, france: totalFrance });
            });
          }
          if (mode === BY_SITE) {
            const demDatas = [];
            data.organizationunits.forEach((dem) => {
              const demToPush = Utils.formatForMonitoringTable(dem);
              demToPush.site = dem.idDem;
              demDatas.push(demToPush);
            });
            const totalFrance = Utils.formatForMonitoringTable(data.France);
            resolve({ dem: demDatas, france: totalFrance });
          }
        });
      });
      if (mode === BY_INTERVIEWER_ONE_SURVEY) {
        Promise.all([p1, p2]).then((data) => {
          cb({
            interviewers,
            site,
            interviewersDetail: data[0],
            total: data[1],
            relevantInterviewers: interviewers,
          });
        });
      }
      if (mode === BY_SITE) {
        p2.then((data) => {
          cb({
            interviewersDetail: data.dem,
            total: data,
          });
        });
      }
    } else {
      p1.then((data) => {
        cb({
          interviewers,
          site,
          interviewersDetail: data,
          relevantInterviewers: interviewers,
        });
      });
    }
  }

  getDataForCampaignPortal(campaignId, cb) {
    const p1 = new Promise((resolve) => {
      this.service.getInterviewersByCampaign(campaignId, (data) => {
        resolve(data);
      });
    });
    const p2 = new Promise((resolve) => {
      this.service.getNotAttributedByCampaign(campaignId, (data) => {
        resolve(data);
      });
    });
    const p3 = new Promise((resolve) => {
      this.service.getTotalDemByCampaign(campaignId, (data) => {
        this.service.getUser((userInfo) => {
          const userOUs = userInfo.localOrganizationUnits.map((x) => x.id);
          const demStateCounts = [];
          data.organizationunits.forEach((dem) => {
            if (userOUs.includes(dem.idDem)) {
              demStateCounts.push({ stateCount: dem });
            }
          });
          const totalDem = Utils.getStateCountSum(demStateCounts);
          resolve(totalDem);
        });
      });
    });
    const p4 = new Promise((resolve) => {
      this.service.getUser((res) => {
        resolve(res.organizationUnit.label);
      });
    });
    Promise.all([p1, p2, p3, p4]).then((data) => {
      cb({
        interviewers: data[0], notAttributed: data[1], total: data[2], site: data[3],
      });
    });
  }
}

export default DataFormatter;
