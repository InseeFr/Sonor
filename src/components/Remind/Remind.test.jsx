// Link.react.test.js
import React from 'react';
import {
  render, screen, fireEvent, cleanup,
} from '@testing-library/react';
import { Router, Route, Switch } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import DataFormatter from '../../utils/DataFormatter';
import Remind from './Remind';
import mocks from '../../tests/mocks';

const history = createMemoryHistory();

const toLocaleDateString = Date.prototype.toLocaleString;
Date.prototype.toLocaleDateString = function() {
  return toLocaleDateString.call(this, 'en-EN', { timeZone: 'UTC',year: "numeric", month: "numeric", day: "numeric" });
};
const OriginalDate = global.Date;
jest
  .spyOn(global, 'Date')
  .mockImplementation((a) => (a ? new OriginalDate(a) : new OriginalDate('2020-08-20T11:01:58.135Z')));
Date.now = jest.fn(() => 1597916474000);

beforeEach(() => {
  history.push('/remind/vqs2021x00');
});

const survey = mocks.surveyVqs;

afterEach(cleanup);

jest.mock('../../utils/DataFormatter');

const mockDataRetreiver = new DataFormatter();

const TestingRouter = ({ ComponentWithRedirection }) => (
  <Router history={history}>
    <Switch>
      <Route path="/remind/vqs2021x00" render={() => <ComponentWithRedirection />} />
      <Route
        path="*"
        render={(routeProps) => (
          <div>
            <div data-testid="Redirect-url">{JSON.stringify(routeProps.history.location.pathname)}</div>
            <div data-testid="Redirect-survey">{!routeProps.history.location || !routeProps.history.location.survey || JSON.stringify(routeProps.history.location.survey)}</div>
          </div>
        )}
      />
    </Switch>
  </Router>
);

it('Component is correctly displayed', async () => {
  const component = render(
    <Router history={history}>
      <Remind location={{ survey }} dataRetreiver={mockDataRetreiver} />
    </Router>,
  );

  // Should match snapshot
  expect(component).toMatchSnapshot();
});

it('Select another survey', async () => {

  const redirectUrl = '/remind/simpsons2020x00';
  const component = render(
    <TestingRouter
      ComponentWithRedirection={
        () => <Remind location={{ survey }} dataRetreiver={mockDataRetreiver} />
      }
    />,
  );

  fireEvent.change(component.getByTestId('Survey_selector'), { target: { value: 'simpsons2020x00' } });

  // Title has changed
  expect(component.baseElement.querySelector('.SurveyTitle').innerHTML).toEqual('Survey on the Simpsons tv show 2020');
});