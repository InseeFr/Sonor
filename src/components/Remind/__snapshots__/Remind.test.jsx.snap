// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component is correctly displayed 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        id="Remind"
      >
        <div
          class="container-fluid"
        >
          <div
            class="row"
          >
            <div
              class="col"
            >
              <a
                class="ButtonLink"
                href="/"
              >
                <button
                  class="YellowButton ReturnButton btn btn-primary"
                  data-testid="return-button"
                  type="button"
                >
                  Back
                </button>
              </a>
            </div>
            <div
              class="col-6"
            >
              <div
                class="SurveyTitle"
              >
                Everyday life and health survey 2021
              </div>
            </div>
            <div
              class="col"
            >
              <form
                class=""
              >
                <div
                  class="form-group"
                >
                  <select
                    class="custom-select custom-select-sm"
                    data-testid="Survey_selector"
                    placeholder="Choose a survey"
                  >
                    <option
                      disabled=""
                      value="-1"
                    >
                      Choose a survey
                    </option>
                    <option
                      value="simpsons2020x00"
                    >
                      Survey on the Simpsons tv show 2020
                    </option>
                  </select>
                </div>
              </form>
            </div>
          </div>
        </div>
        <div
          class="ViewCard card"
        >
          <div
            class="card-title h5"
          >
            Page in development
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      id="Remind"
    >
      <div
        class="container-fluid"
      >
        <div
          class="row"
        >
          <div
            class="col"
          >
            <a
              class="ButtonLink"
              href="/"
            >
              <button
                class="YellowButton ReturnButton btn btn-primary"
                data-testid="return-button"
                type="button"
              >
                Back
              </button>
            </a>
          </div>
          <div
            class="col-6"
          >
            <div
              class="SurveyTitle"
            >
              Everyday life and health survey 2021
            </div>
          </div>
          <div
            class="col"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <select
                  class="custom-select custom-select-sm"
                  data-testid="Survey_selector"
                  placeholder="Choose a survey"
                >
                  <option
                    disabled=""
                    value="-1"
                  >
                    Choose a survey
                  </option>
                  <option
                    value="simpsons2020x00"
                  >
                    Survey on the Simpsons tv show 2020
                  </option>
                </select>
              </div>
            </form>
          </div>
        </div>
      </div>
      <div
        class="ViewCard card"
      >
        <div
          class="card-title h5"
        >
          Page in development
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
